sensor: adding a driver for the STM32 quadrature encoder

valeriosetti opened the pull request:
This PR adds a driver for the STM32 internal quadrature encoder and provides a sample application for its usage.

The driver is quite simple since and it's strongly inspired by:
- "drivers/sensor/qdec_sam" for the sensor part related to a quadrature encoder
- "drivers/counter/counter_ll_stm32_timer.c" for the ST part (interfacing with the parent timer at DT level)

valeriosetti committed:
samples: sensor: adding a sample application for the STM32 qdec driver

Sample application for the STM32 quadrature encoder driver

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

erwango commented on drivers/sensor/qdec_stm32/qdec_stm32.c:
```diff
@@ -0,0 +1,146 @@
+		return retval;
+	}
+
+	retval = clock_control_on(DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE),
```
Check device is ready before using. See https://github.com/zephyrproject-rtos/zephyr/pull/48787

erwango commented on drivers/sensor/qdec_stm32/qdec_stm32.c:
```diff
@@ -0,0 +1,146 @@
+		init_props.IC2ActiveInput = LL_TIM_IC_POLARITY_FALLING;
+	}
+
+	init_props.IC1Filter = dev_cfg->input_filtering_level *
+								LL_TIM_IC_FILTER_FDIV1_N2;
+	init_props.IC2Filter = dev_cfg->input_filtering_level *
+								LL_TIM_IC_FILTER_FDIV1_N2;
```
These should fit in one line

valeriosetti replied:
If I try to put the code in 1 line then it exceeds 80 characters... Or am I missing something in your request?

erwango replied:
80 is not a hard limit anymore, 100 is the new 80.
Besides, checkpatch counts tab as 8 chars, so your 8 tabs would count as 64 chars here, so this line is > 80 anyway.

erwango commented on dts/arm/st/f4/stm32f4.dtsi:
```diff
@@ -428,6 +459,12 @@
+				status = "disabled";
+				input-filter-level = <0>;
+			};
 		};
```
It would be nice to at least complete the addition of qdec nodes for the f4 series (f405.dtsi, f410.dtsi, f412.dtsi)

valeriosetti replied:
As suggested by @str4t0m I looked more in detail on the Reference Manuals of F405, F410 and F412 and found that timers which support this quadrature encoder feature are: 1 to 5, 8.
I'll fix the code accordingly

erwango commented on dts/bindings/sensor/st,stm32-qdec.yaml:
```diff
@@ -0,0 +1,45 @@
+     input-filter-level:
+       type: int
+       required: true
+       description: intensity of filtering on the input
+       default: 0
+       enum:
+         - 0  # No filter
+         - 1  # FDIV1_N2
+         - 2  # FDIV1_N4
+         - 3  # FDIV1_N8
+         - 4  # FDIV2_N6
+         - 5  # FDIV2_N8
+         - 6  # FDIV4_N6
+         - 7  # FDIV4_N8
+         - 8  # FDIV8_N6
+         - 9  # FDIV8_N8
+         - 10 # FDIV16_N5
+         - 11 # FDIV16_N6
+         - 12 # FDIV16_N8
+         - 13 # FDIV32_N5
+         - 14 # FDIV32_N6
+         - 15 # FDIV32_N8
```
This could use a better description. Also, please note that comments are not rendered in zephyr docs.
See https://docs.zephyrproject.org/latest/build/dts/api/bindings/counter/st%2Cstm32-counter.html#dtbinding-st-stm32-counter.

erwango commented on samples/sensor/stm32_qdec/prj.conf:
```diff
@@ -0,0 +1,4 @@
+CONFIG_SENSOR=y
+CONFIG_PRINTK=y
+CONFIG_CBPRINTF_FP_SUPPORT=y
+CONFIG_SENSOR=y
```
Duplicate

erwango commented on samples/sensor/stm32_qdec/README.rst:
```diff
@@ -0,0 +1,62 @@
+.. _stm32_qdec_sensor:
+
+STM32 Quadrature Encoder Sensor
```
Shouldn't it be rather 'decoder' than 'encoder' ?

valeriosetti replied:
This is something I also wasn't sure about.
I followed similar drivers' design and since all of them had "qdec" in the name (which I guess stands for "quadrature decoder") I did the same. However ST in their documentation talks about "quadrature encoder" (see RM0368 for example), so I don't know which of the 2 should prevail...

valeriosetti replied:
Actually this has been solved following a suggestion from @str4t0m. Now the sample application is more generic and should be able to handle also decoders from other manufacturers. As a consequence I changed the name to "decoder"

erwango commented on samples/sensor/stm32_qdec/README.rst:
```diff
@@ -0,0 +1,62 @@
+
+In this example:
+
+* ``timers3`` is used as implementation, but it can be replaced with other
```
```suggestion
* ``timers3`` is used as example, but it can be replaced with other
```

erwango commented on samples/sensor/stm32_qdec/boards/nucleo_f401re.overlay:
```diff
@@ -0,0 +1,20 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Application overlay for creating vbat sensor device instance
```
```suggestion
 * Application overlay for creating quadrature decoder device instance
```

erwango reviewed (changes requested):
Thanks for this contribution.
Some changes required but overall looks good.
Also, please split first commits into dts and driver parts

valeriosetti committed:
drivers: sensor: adding driver for STM32 quadrature encoder

This driver configures STM32's internal timers in order to use
the quadrature encoder mode.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
dts: sensor: adding nodes for STM32 quadrature encoder

Nodes for quadrature encoders are added to the STM32F4 devices

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

str4t0m reviewed (changes requested):
A few comments.

valeriosetti commented on dts/arm/st/f4/stm32f4.dtsi:
```diff
@@ -323,6 +330,12 @@
+				input-filter-level = <0>;
+			};
 		};
 
 		timers4: timers@40000800 {
```
I apologize for the mistake.
After your comment I looked more in detail on the Reference Manuals of the F4's family and found that timers which support this quadrature encoder feature are: 1 to 5, 8.
I'll fix the code accordingly.

valeriosetti commented on dts/arm/st/f4/stm32f4.dtsi:
```diff
@@ -407,6 +432,12 @@
 				compatible = "st,stm32-counter";
 				status = "disabled";
 			};
+
+			qdec {
+				compatible = "st,stm32-qdec";
```
You're right, I did a mistake here. This was fixed as specified in the previous comment

valeriosetti commented on samples/sensor/stm32_qdec/README.rst:
```diff
@@ -0,0 +1,62 @@
+.. _stm32_qdec_sensor:
+
+STM32 Quadrature Encoder Sensor
+###############################
```
You're right, this would be nice! I'll try to implement it

valeriosetti commented on drivers/sensor/qdec_stm32/qdec_stm32.c:
```diff
@@ -0,0 +1,146 @@
+{
+	struct qdec_stm32_dev_data *const dev_data = dev->data;
+
+	if (chan == SENSOR_CHAN_ROTATION) {
```
Unfortunately I'm not sure I can determine how much the physical device rotated in degrees once I get a certain value of the counter. I mean, the number of edges for an entire rotation might change between various mechanical encoders.
Perhaps adding another identifier of raw type would be better?

gmarull replied:
If sensor supports `SENSOR_CHAN_ROTATION` it should return the value in the specified units, ie degrees. A custom/new channel can be introduced to report in "counts" if needed. The driver may support an attribute to set the counts/revolution ratio so that it can support the standard `SENSOR_CHAN_ROTATION`. If other drivers are not doing this, they should really be fixed as it is a bug.

erwango commented on drivers/sensor/qdec_stm32/Kconfig:
```diff
@@ -0,0 +1,12 @@
+config QDEC_STM32
+	bool "STM32 QDEC driver"
+	default y
+	depends on SOC_FAMILY_STM32
```
Not required, please remove

erwango commented on drivers/sensor/qdec_stm32/qdec_stm32.c:
```diff
@@ -0,0 +1,155 @@
+			.enr = DT_CLOCKS_CELL(DT_INST_PARENT(n), bits)			  \
+		},		\
+		.prescaler = DT_PROP(DT_INST_PARENT(n), st_prescaler), \
+		.is_input_polarity_inverted = DT_INST_PROP(n,		\
+								input_polarity_inverted),	\
```
Ditto on line length

erwango reviewed (changes requested):
Please rename sample (`samples/sensor/stm32_qdec` > `samples/sensor/qdec`)

gmarull commented on samples/sensor/qdec/sample.yaml:
```diff
@@ -0,0 +1,15 @@
+tests:
+  sample.sensor.qdec_sensor:
+    tags: sensors
+    platform_allow: nucleo_f401re
```
samples under `samples/sensor` should not be board specific, please, filter based on a particular alias being enabled in this case. You can see the generic accelerometer sample as a reference.

erwango replied:
In case of generic accelerometer sample, we have in tree boards available with required hardware (accel sensors).
This won't be the case for current sensor and providing qdec configuration on random boards as part of their default configuration doesn't follow the board description guidelines (and first of all doesn't feel right anyway).
Current approach of using an overlay seems like a better approach. 

The unfortunate consequence is the restriction of the sample to boards for which an overlay is provided, as filtering based on a particular alias being enabled in this case is not possible in the case of configuration made using sample overlays (which is a "limitation" in the way twister applies filter).

I understand your request, but I don't see how to solve it easily.

gmarull replied:
right, I always forget about this limitation in twister...

gmarull commented on samples/sensor/qdec/src/main.c:
```diff
@@ -0,0 +1,46 @@
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/sys/printk.h>
+
+#if !DT_NODE_HAS_STATUS(DT_ALIAS(qdec0), okay)
+#error "Unsupported board: qdec0 devicetree alias is not defined"
+#endif
```
this is redundant, `DEVICE_DT_GET` will fail anyway

gmarull commented on samples/sensor/qdec/sample.yaml:
```diff
@@ -0,0 +1,15 @@
+    tags: sensors
+    platform_allow: nucleo_f401re
+    timeout: 10
+    harness: console
```
because it requires to connect an external device, harness can't be console per my understanding cc: @PerMac

valeriosetti replied:
well actually the external device is just a mechanical encoder, so it's not strictly required to initialize the internal timer and run the example. Of course it's mandatory if the user wants to see counter's value changing over time.
Btw, since your point is correct and this feature is hard to auto-test, I changed this part limiting it to build only

PerMac replied:
Harness: console means that twister will look for a given regexes in the console output. If the platform prints a required log in the output, harness console is fine. If an extra hardware is required (e.g. sensor connected, shield or some pin being wired) one needs to define a required `fixture` and twister will only use platforms with has such modification. I saw this is resolved, but since I was pinged I added my 2 cents. https://docs.zephyrproject.org/latest/develop/test/twister.html#fixtures

gmarull commented on samples/sensor/qdec/sample.yaml:
```diff
@@ -0,0 +1,15 @@
+      ordered: true
+      regex:
+        - "Quadrature decoder sensor test"
+        - "Counter value = 0"
```
is it guaranteed to always start with zero?

valeriosetti replied:
At least on ST devices the answer should be yes, but since I agree this is not general and also according to the previous comment, I removed this part

gmarull commented on dts/bindings/sensor/st,stm32-qdec.yaml:
```diff
@@ -0,0 +1,66 @@
+         13: Fs = F_clk/32, N=5
+         14: Fs = F_clk/32, N=6
+         15: Fs = F_clk/32, N=8
+       default: 0
```
defaults need to be justified per dt guidelines

valeriosetti replied:
At least on ST devices the answer should be yes, but since I agree this is not general and also according to the previous comment, I removed this part

MaureenHelm replied:
There is an explanation of what the default value is, but not why it was chosen.

gmarull commented on drivers/sensor/qdec_stm32/qdec_stm32.c:
```diff
@@ -0,0 +1,152 @@
+#include <zephyr/sys/util.h>
+#include <zephyr/device.h>
+#include <zephyr/init.h>
+#include <soc.h>
```
why is `<soc.h>` needed?

valeriosetti replied:
sorry, that was a leftover from the reference driver I started from

gmarull reviewed (changes requested):
a few nitpicks

gmarull commented on drivers/sensor/qdec_stm32/qdec_stm32.c:
```diff
@@ -0,0 +1,151 @@
+	struct stm32_pclken pclken;
+	TIM_TypeDef *timer_inst;
+	uint16_t prescaler;
+	uint8_t is_input_polarity_inverted;
```
`bool`

gmarull commented on drivers/sensor/qdec_stm32/qdec_stm32.c:
```diff
@@ -0,0 +1,151 @@
+	init_props.IC2Filter = dev_cfg->input_filtering_level * LL_TIM_IC_FILTER_FDIV1_N2;
+
+	if (IS_TIM_32B_COUNTER_INSTANCE(dev_cfg->timer_inst)) {
+		LL_TIM_SetAutoReload(dev_cfg->timer_inst, 0xFFFFFFFF);
```
nit: `UINT32_MAX`?

gmarull commented on drivers/sensor/qdec_stm32/qdec_stm32.c:
```diff
@@ -0,0 +1,151 @@
+	if (IS_TIM_32B_COUNTER_INSTANCE(dev_cfg->timer_inst)) {
+		LL_TIM_SetAutoReload(dev_cfg->timer_inst, 0xFFFFFFFF);
+	} else {
+		LL_TIM_SetAutoReload(dev_cfg->timer_inst, 0xFFFF);
```
`UINT16_MAX`?

gmarull commented on drivers/sensor/qdec_stm32/qdec_stm32.c:
```diff
@@ -0,0 +1,151 @@
+
+	LL_TIM_EnableCounter(dev_cfg->timer_inst);
+
+	LOG_INF("Device %s initialized", dev->name);
```
`LOG_DBG`? is it actually worth?

valeriosetti replied:
it was present in the driver I used as reference, so I kept it, but you're right. I can remove it

gmarull commented on dts/arm/st/f4/stm32f4.dtsi:
```diff
@@ -281,6 +283,12 @@
+			qdec {
+				compatible = "st,stm32-qdec";
+				status = "disabled";
+				input-filter-level = <NO_FILTER>;
```
should property be prefixed with `st,`? @erwango

erwango replied:
Exact, thanks for noticing.

PerMac commented on samples/sensor/qdec/sample.yaml:
```diff
@@ -0,0 +1,8 @@
+  sample.sensor.qdec_sensor:
+    tags: sensors
+    platform_allow: nucleo_f401re
+    build_only: true
```
with build_olny: true you won't be able to test it with twister at all, even if you have the required setup. I assume the sample needs an extra sensor connected to a platform. If so, you can use fixture keyword, and then twister will only allow to execute this sample if a proper setup (with a fixture) is defined https://docs.zephyrproject.org/latest/develop/test/twister.html#fixtures If not, it will just build the sample on a given platform.
From my understanding we require samples to also define a test which can validate them, if possible. If the sample provides some output, based on which a user is able to validate if it worked (e.g. "sensor active, readout XXX") this can be automated in twister, using console: harness and defining required regexes

erwango commented:
@valeriosetti Can you add yourself in front of this driver in CODEOWNERS file ?

valeriosetti committed:
CODEOWNERS: add ownership of STM32 quadrature decoder driver

Adding myself as the code owner for the STM32's quadrature encoder
driver

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

gmarull commented on drivers/sensor/qdec_stm32/qdec_stm32.c:
```diff
@@ -0,0 +1,166 @@
+
+static int qdec_stm32_initialize(const struct device *dev)
+{
+	__ASSERT_NO_MSG(dev != NULL);
```
when is this condition possible?

valeriosetti replied:
Sorry, even this one was a leftover from the driver I started from. I'll remove it

gmarull commented on dts/arm/st/f4/stm32f4.dtsi:
```diff
@@ -281,6 +283,13 @@
+				compatible = "st,stm32-qdec";
+				status = "disabled";
+				st,input-filter-level = <NO_FILTER>;
+				st,counts-per-revolution = <10>;
```
why is this set at dts/ level? Seems like a mechanical property of the encoder, right?

valeriosetti replied:
so, if I understand correctly, you say it's better to keep this on the board's DTS/overlay only, right?

gmarull replied:
What I'd do:

- `st,input-filter-level`: set its default in the binding to `NO_FILTER` (justification: SoC default). Adding it to `dts/` it's de-facto setting a default. If not, force end user (ie board) to make the choice
- `st,counts-per-revolution`: force end user since it's a mechanical property of the encoder they'll use

gmarull commented on dts/bindings/sensor/st,stm32-qdec.yaml:
```diff
@@ -0,0 +1,76 @@
+     st,counts-per-revolution:
+       type: int
+       required: true
+       default: 10
+       description: |
+         This is a number >= 1 that is used to determine how many revolutions
+         were done based on the current counter's value.
+         The default value is 1 just for example, but it needs to be adjusted
+         to the physical encoder used.
```
default/description mismatch. This property should be required, since it depends on the encoder connected to the board.

valeriosetti replied:
I apologize sorry for the typo. And you're right, it should also be required

mbolivar-nordic reviewed (commented):
Binding LGTM

carlescufi commented:
@gmarull @erwango please take a look

PerMac commented:
I left a comment regarding `build_only: true` for the `sample.yaml` earlier on and I think it was not addressed. I think it is a valid one: currently this sample is untestable due to this parameter and lack of expected output. I proposed to use fixture, which IMO would do the trick, allowing testing but only on a properly marked setups (assuming the expected output is also provided). More in that comment.  I refrenie from adding +/- review since I know nothing about the driver in a PR and if its testing is feasible.

erwango commented on samples/sensor/qdec/sample.yaml:
```diff
@@ -0,0 +1,16 @@
+tests:
+  sample.sensor.qdec_sensor:
+    tags: sensors
+    filter: dt_alias_exists("qdec0")
```
FYI, this won't work to build sample on nucleo_f401re as overlays are applied after filtering.
This could be useful for your own CI set up with a board that defines this alias as default configuration locally.
But in current state of the tree with no board defining it; result is that the sample will never be built.
Besides, I don't encourage adding this alias directly as part of nucleo_f401re default description as the required associated encoder hw is not part of the BSP, so it would not make much sense.

So, I'd suggest to keep `platform_allow: nucleo_f401re`.

valeriosetti replied:
Thanks! I didn't know that overlays are applied only after filtering. I'll fix this as you suggested

str4t0m commented on samples/sensor/qdec/README.rst:
```diff
@@ -0,0 +1,70 @@
+
+.. code-block:: console
+
+    Counter value = 0
+    Counter value = 1
+    Counter value = 2
```
This prints do not match the sample anymore.

gmarull replied:
needs to be updated as well

str4t0m commented on samples/sensor/qdec/README.rst:
```diff
@@ -0,0 +1,70 @@
+        status = "okay";
+        st,prescaler = <3>;
+
+        qdec: qdec {
+            status = "okay";
+            pinctrl-0 = <&tim3_ch1_pa6 &tim3_ch2_pa7>;
+            pinctrl-names = "default";
+            input-polarity-inverted;
+            input-filter-level = <15>;
```
If an example is given it should match the current bindings.
Could you update the `input-filter-level` to match the actual overlay and add the `st,counts-per-revolution`

gmarull replied:
this needs to be updated to match latest prop names/defines

str4t0m commented on dts/bindings/sensor/st,stm32-qdec.yaml:
```diff
@@ -0,0 +1,73 @@
+       required: false
+       description: Encoder is triggered by a falling edge on the input pin
+
+     st,input-filter-level:
+       type: int
+       required: true
+       description: |
+         Intensity of the filter applied to the input signal. This is
+         implemented by scaling the sampling frequency and adding a counter
+         in which N consecutive samples with same value are needed to validate
+         a transition.
+         Mapping is as follows (F_clk is the timer's clock):
+         0: No filter, sampling is done at F_dts (default value)
```
It does not make sense to define a default value and mark the property as required.
Please choose either or.

gmarull commented on samples/sensor/qdec/src/main.c:
```diff
@@ -0,0 +1,42 @@
+			return;
+		}
+
+		printk("Value = %d", val.val1);
```
```suggestion
		printk("Position = %d degrees", val.val1);
```

valeriosetti replied:
At first I was going to write it the same way you proposed, but then I thought that so far not all the quadrature decoder drivers return an angle. For example I don't think the "drivers/sensor/qdec_sam does; it seems to me that it just returns the raw counter value. Therefore I was not sure of what was the best solution here...
Should I assume that from now on all the decoder driver will return an angle?

valeriosetti replied:
@gmarull 
I did some other check on the other qdec drivers and I found that the behavior is not really homogeneous:
- "drivers/sensor/qdec_sam/" --> returns the raw value of the counter on val1 and nothing on val2
- "drivers/sensor/qdec_nrfx/" --> returns a differential value (compared to the previous read) with an integer number of turns on val1 and the remainder on val2
- mine, according to the required changes of this PR, returns the absolute angle position on val1 and nothing on val2

I don't know if the change you suggested applies well here... What do you think?

gmarull replied:
Yup, that's indeed a problem that needs to be resolved. If a driver supports the rotation channel, it should provide value in degrees, not counts. Driver could support 2 channels, though, one for counts and another for degrees. About using val1/val2: it's all about the resolution driver supports. When using counts, val2 doesn't make sense, but when using degrees it may be used if you have a resolution < 1 degree.

gmarull commented on samples/sensor/qdec/src/main.c:
```diff
@@ -0,0 +1,42 @@
+{
+	struct sensor_value val;
+	int rc;
+	const struct device *dev = DEVICE_DT_GET(DT_ALIAS(qdec0));
```
```suggestion
	const struct device *const dev = DEVICE_DT_GET(DT_ALIAS(qdec0));
```

gmarull commented on samples/sensor/qdec/src/main.c:
```diff
@@ -0,0 +1,42 @@
+	const struct device *dev = DEVICE_DT_GET(DT_ALIAS(qdec0));
+
+	if (!device_is_ready(dev)) {
+		printk("Qdec device is not ready\n");
```
```suggestion
		printk("Encoder device is not ready\n");
```

gmarull commented on samples/sensor/qdec/src/main.c:
```diff
@@ -0,0 +1,42 @@
+			return;
+		}
+
+		rc = sensor_channel_get(dev, SENSOR_CHAN_ROTATION, &val);
```
note: this sample is now generic to any kind of sensor that is able to report a rotation channel in degrees, qdec devices are a typical case but we could have others. Maybe rename sample folder?

valeriosetti replied:
Ok, I understand your point, but I don't have any good idea about a possible naming for such a generic folder. Do you have any suggestion as well? What about "rotation"?

PS = increasing the scope of this sample application will cause changes also to the alias used in the DT, since `qdec` will no more be a good choice

gmarull replied:
Let's keep the `qdec` naming for now, fair enough.

gmarull commented on samples/sensor/qdec/src/main.c:
```diff
@@ -0,0 +1,42 @@
+			return;
+		}
+
+		rc = sensor_channel_get(dev, SENSOR_CHAN_ROTATION, &val);
```
note: this sample is now generic to any kind of sensor that is able to report a rotation channel in degrees, qdec devices are a typical case but we could have others. Maybe rename sample folder?

gmarull commented on samples/sensor/qdec/src/main.c:
```diff
@@ -0,0 +1,42 @@
+		return;
+	}
+
+	printk("Quadrature decoder sensor test\n");
```
see my comment below

valeriosetti replied:
Yes, once we solve that comment, I'll set this accordingly

gmarull commented on drivers/sensor/qdec_stm32/qdec_stm32.c:
```diff
@@ -0,0 +1,165 @@
+ * @brief STM32 family Quadrature Decoder (QDEC) driver.
+ */
+
+#include <errno.h>
+#include <zephyr/sys/__assert.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/device.h>
+#include <zephyr/init.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/clock_control/stm32_clock_control.h>
+
+#include <stm32_ll_tim.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(qdec_stm32, CONFIG_SENSOR_LOG_LEVEL);
```
```suggestion
#include <errno.h>

#include <zephyr/sys/__assert.h>
#include <zephyr/sys/util.h>
#include <zephyr/device.h>
#include <zephyr/init.h>
#include <zephyr/drivers/sensor.h>
#include <zephyr/drivers/pinctrl.h>
#include <zephyr/drivers/clock_control/stm32_clock_control.h>
#include <zephyr/logging/log.h>

#include <stm32_ll_tim.h>

LOG_MODULE_REGISTER(qdec_stm32, CONFIG_SENSOR_LOG_LEVEL);
```

gmarull reviewed (commented):
last few nitpicks

valeriosetti commented:
**Note about this last forced push**
I removed the use of timer's prescaler because while extensively testing I found it could cause wrong behavior of the encoder.
- When prescaler is 0, so that it's basically disabled, everything works as expected
- On the other hand, once it's set to something >= 1 (therefore it's enabled), the counter works normally as long as it is upcounting; but if I turn the knob the other way in order to cause an underflow (in which case the counter should be loaded with the max value), actually it remains stuck at 0 forever

Btw ``st,counts-per-revolution`` can still be used to scale the returned value in software
